### PR TITLE
Add std/esm benchmark.

### PR DESCRIPTION
**Pull request by jdalton**
This adds a basic [`@std/esm`](https://github.com/standard-things/esm) benchmark of loading 643 `lodash-es` modules. The `@std/esm` module loader will be used to power the next major release of [lodash](https://gist.github.com/anvaka/8e8fa57c7ee1350e3491#file-01-most-dependent-upon-md).

It has a bunch of moving parts. On first load it does a parse using a highly tweaked version of acorn. On subsequent loads it simulates the ESM runtime with its own module loading pipeline.

This test is for the cli runs and should be skipped for browser runs.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -34,6 +34,7 @@
   "license": "BSD-3-Clause",
   "dependencies": {
     "@babel/standalone": "7.0.0-beta.32",
+    "@std/esm": "^0.21.1",
```
Comment (bmeurer) [inline package.json]: We use exact dependencies.

File: package.json [hunk 2]
```diff
@@ -45,6 +46,7 @@
     "esprima": "4.0.0",
     "jshint": "2.9.5",
     "lebab": "2.7.7",
+    "lodash-es": "^4.17.5",
```
Comment (bmeurer) [inline package.json]: Same here. Do we even need the explicit `lodash-es` dependency?

File: .esmrc [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (bmeurer) [inline .esmrc]: This file is not bundled with the webpack build? Can we avoid external configuration files?

File: package.json (hunk 2 above)
Comment (jdalton) [reply package.json]: Would committing the entire package to the vendor folder be a better place?

File: package.json (hunk 1 above)
Comment (jdalton) [reply package.json]: > Thanks for the pull request. There's one thing I'm wondering about though: Does @std/esm count as a web developer tool? And does it even make sense to consider outside of Node?

Not outside Node no. It's Node for sure. Is there a more appropriate suite?

Comment (jdalton) [reply package.json]: The suite description states
> This is a benchmark suite designed to measure the JavaScript related workloads commonly used by Web Developers nowadays, for example the core workloads in popular tools like Babel or TypeScript.

Lodash is on more than 2 million websites and impacts over 150,000 npm packages (many of those used by web devs). Seems like a good fit.

Comment (bmeurer) [reply package.json]: Yeah, good point. I do think we should have a `lodash` benchmark in there for sure. `lodash` is one of the most popular `npm` (frontend) dependencies (if not the most popular).

As for `@std/esm`, we should probably create a Node tooling benchmark for Node only/mostly packages. WDYT?

Comment (bmeurer) [reply package.json]: Do you have a good `lodash` benchmark suite already?

Comment (jdalton) [reply package.json]: > Do you have a good lodash benchmark suite already?

I have a pretty robust one. It allows swapping out different versions of Lodash/Underscore. It's a long running benchmark though designed to ensure the things Lodash is good at stay good and the things its bad at stay reasonable.

> As for @std/esm, we should probably create a Node tooling benchmark for Node only/mostly packages. WDYT

That's cool too though so testing the Node flavors of Babel, Webpack, etc.?

Comment (bmeurer) [reply package.json]: I'm very interested in the `lodash` benchmark. As long as the runtime is roughly comparable to the other tests it should be fine, and if not we could reduce the iteration count a bit I guess.

> That's cool too though so testing the Node flavors of Babel, Webpack, etc.?

I don't think it makes sense to have webpack there. This is definitely something that should be owned and operated by the [benchmarking WG](https://benchmarking.nodejs.org/). Maybe you can join the call on Monday to discuss the idea?

Comment (jdalton) [reply package.json]: > As long as the runtime is roughly comparable to the other tests it should be fine, and if not we could reduce the iteration count a bit I guess.

It'd need some tweaking it's a [pretty massive suite](https://github.com/lodash/lodash/blob/4.17.5/perf/perf.js).

> This is definitely something that should be owned and operated by the benchmarking WG. Maybe you can join the call on Monday to discuss the idea?

Sure thing! You can DM details if you'd like.

Comment (bmeurer) [reply package.json]: > It'd need some tweaking it's a pretty massive suite.

It also seems to mess with the global state of benchmark.js, which needs to be fixed.

> Sure thing! You can DM details if you'd like.

Added note to https://github.com/nodejs/benchmarking/issues/198 for monday's meeting.

**Comment by jdalton** [comment]
Cool. I'll close this then to go down the WG route.

**Comment by bmeurer** [comment]
Awesome, thanks. Looking forward to a `lodash` pull request 👍

**Comment by mathiasbynens** [comment]
I’d love to hear more about why you think adding a lodash benchmark to WTB makes sense. 

lodash is an *immensely* popular and important project for the ecosystem. It’s not, however, a “tool” that’s used in build scripts (at least not directly) in the same way that acorn, babel, babylon, buble, chai, coffeescript, espree, esprima, jshint, lebab, prepack, prettier, source-map, typescript, uglify-es, uglify-js are.

lodash fits better in the category of frameworks/libraries IMHO, and therefore I’m thinking perhaps Speedometer would be a more appropriate benchmark in which to include lodash. WDYT?

I worry that adding a lodash benchmark to WTB would not necessarily represent its real-world usage in web developer tools — it would be more akin to a synthetic benchmark. Am I looking at this the wrong way?

**Comment by jdalton** [comment]
> Awesome, thanks. Looking forward to a lodash pull request 👍

Okay, will do too!

**Comment by bmeurer** [comment]
I was thinking of `lodash` as a tool in the developers toolbox, so from that point of view it made sense to include it here. But I see your point that it'd be a synthetic benchmark if include `lodash` itself. It's definitely more on the framework/library side. Having a tool in the benchmark suite that's heavily based on lodash makes sense, though. Is there one?

Besides that, I guess `lodash` should be in Speedometer then. IIRC there's already one or two tests that make heavy use of underscore.

**Comment by bmeurer** [comment]
Resolution from benchmarking WG meeting: We will create a separate benchmark suite with the Node specific tools/libraries/loaders/runtimes, and it'll live in https://github.com/nodejs/benchmarking. @jdalton signed up for this.

**Pull request closed without merging**